internal/resourcepolicies/resource_policies.go

```diff
@@ -0,0 +1,109 @@
+}
+
+// VolumePolicy defined policy to conditions to match Volumes and related action to handle matched Volumes
+type VolumePolicy struct {
+	// Conditions defined list of conditions to match Volumes
+	Conditions map[string]interface{} `yaml:"conditions"`
+	Action     Action                 `yaml:"action"`
+}
+
+// ResourcePolicies currently defined slice of volume policies to handle backup
+type ResourcePolicies struct {
+	Version        string         `yaml:"version"`
+	VolumePolicies []VolumePolicy `yaml:"volumePolicies"`
+	// we may support other resource policies in the future, and they could be added separately
+	// OtherResourcePolicies: []OtherResourcePolicy
+}
+
+type ResourcePolicyMatcher interface {
+	Match(res interface{}, actionType string) bool
+}
+
+type VolumePolicyMatcherImpl struct {
+	policy      VolumePolicy
+	matcherFunc func(volume interface{}, actionType string) bool
+}
+
+func (vpm *VolumePolicyMatcherImpl) Match(volume interface{}, actionType string) bool {
+	return vpm.matcherFunc(volume, actionType)
+}
+
+// Match interface is implemented by VolumePolicy
+func (policy *VolumePolicy) Match(volume interface{}, actionType string) bool {
+	if policy.Action.Type != actionType {
+		return false
+	}
+
+	val, ok := volume.(*VolumeConditions)
+	if !ok {
+		return false
+	}
+
+	if volConditions, err := MarshalVolumeConditions(policy.Conditions); err != nil {
+		return false
+	} else {
+		return val.IsConditionsMatch(volConditions)
+	}
+}
+
+func NewVolumePolicyMatcher(policy VolumePolicy) ResourcePolicyMatcher {
+	return &VolumePolicyMatcherImpl{
+		policy:      policy,
+		matcherFunc: policy.Match,
+	}
+}
+
+type ResourcePoliciesMatcherFactory struct {
+	volumeMatchers []ResourcePolicyMatcher
+	// could extended by add other policy matcher
+	// otherResourceMatchers   []ResourcePolicyMatcher
+}
+
+func (r *ResourcePoliciesMatcherFactory) AddVolumePolicyMatcher(volumePolicy []VolumePolicy) {
+	for _, policy := range volumePolicy {
+		matcher := NewVolumePolicyMatcher(policy)
+		r.volumeMatchers = append(r.volumeMatchers, matcher)
+	}
+}
+
+func NewResourcePoliciesMatcherFactory(policies *ResourcePolicies) *ResourcePoliciesMatcherFactory {
+	factory := &ResourcePoliciesMatcherFactory{}
+	factory.AddVolumePolicyMatcher(policies.VolumePolicies)
+	// TODO we can add other policy matcher into the factory
+	// factory.AddOtherResourcePolicyMatcher(policies.OtherResourcePolicies)
+	return factory
+}
+
+// GetMatchers return related resource policy matcher
+func (rpf *ResourcePoliciesMatcherFactory) GetMatchers(resourceType string) []ResourcePolicyMatcher {
+	switch resourceType {
+	case "volume":
+		return rpf.volumeMatchers
+	default:
+		return nil
+	}
+}
+
+// IsVolumeMatch checks the current volume is match resource policies
+// It will return once matched ignoring the latter policies
+func IsVolumeMatch(policies *ResourcePolicies, volume *VolumeConditions, actionType string) bool {
```

Comment:
If we do it this way, we end up doing repeated checks b/c when there are more actions needed you may want to call
`IsVolumeMatch(p, v, "skip")`
`IsVolumeMatch(p,v,"volume-snapshot")` ....

How about we do it like 
`MatchVolume(p, v) *Action`
so we only go through the matchers once and find the action for the volume?

Reply:
sounds good